hal/src/nRF52840/sleep_hal.cpp

```diff
@@ -260,6 +289,84 @@ static void configRtcWakeupSource(const hal_wakeup_source_base_t* wakeupSources)
             nrf_rtc_event_clear(NRF_RTC2, NRF_RTC_EVENT_COMPARE_0);
             nrf_rtc_int_enable(NRF_RTC2, NRF_RTC_INT_COMPARE0_MASK);
             nrf_rtc_event_enable(NRF_RTC2, RTC_EVTEN_COMPARE0_Msk);
+            break; // Stop traversing the list.
+        }
+        source = source->next;
+    }
+}
+
+static void configLpcompWakeupSource(const hal_wakeup_source_base_t* wakeupSources) {
+    Hal_Pin_Info* halPinMap = HAL_Pin_Map();
+    auto source = wakeupSources;
+    while (source) {
+        if (source->type == HAL_WAKEUP_SOURCE_TYPE_LPCOMP) {
+            auto lpcompWakeup = reinterpret_cast<const hal_wakeup_source_lpcomp_t*>(source);
+            nrf_lpcomp_config_t config = {};
+
+            NVIC_DisableIRQ(COMP_LPCOMP_IRQn);
+            NVIC_ClearPendingIRQ(COMP_LPCOMP_IRQn);
+
+            // Reference voltage is VDD/16 * N, N ranges from 1 to 15, VDD/16 = 206mV
+            constexpr uint8_t vddDiv16 = 206;
+            constexpr nrf_lpcomp_ref_t refs[] = {
+                NRF_LPCOMP_REF_SUPPLY_1_16, NRF_LPCOMP_REF_SUPPLY_1_8,
+                NRF_LPCOMP_REF_SUPPLY_3_16, NRF_LPCOMP_REF_SUPPLY_2_8,
+                NRF_LPCOMP_REF_SUPPLY_5_16, NRF_LPCOMP_REF_SUPPLY_3_8,
+                NRF_LPCOMP_REF_SUPPLY_7_16, NRF_LPCOMP_REF_SUPPLY_4_8,
+                NRF_LPCOMP_REF_SUPPLY_9_16, NRF_LPCOMP_REF_SUPPLY_5_8,
+                NRF_LPCOMP_REF_SUPPLY_11_16, NRF_LPCOMP_REF_SUPPLY_6_8,
+                NRF_LPCOMP_REF_SUPPLY_13_16, NRF_LPCOMP_REF_SUPPLY_7_8,
+                NRF_LPCOMP_REF_SUPPLY_15_16
+            };
+            constexpr uint8_t elements = sizeof(refs) / sizeof(nrf_lpcomp_ref_t);
+            uint8_t n = (lpcompWakeup->voltage + (vddDiv16 >> 1/* vddDiv16 / 2 */)) / vddDiv16;
```

Comment:
Nit: the compiler is smart enough to do bitshifts and any other optimizations on its own.